# Week 1 Tutorial - Data Types, Data Structures, and Basic Operations

For the first class, we will cover the following topics:

1. Python Programming in Biology 
2. Using Jupyter Notebook
3. Data Types and Basic Operations 
    * Integers, floating-point numbers, and arithmetic operations
    * Strings and string operations
    * Comparison operations
    * Booleans and logical operations 
    * Lists, tuples, dictionaries, and sets 
    * Indexing and slicing of sequence data types 
    * Subscriptable, mutable, and iterable data types 
6. Output and Input 
    * print()
    * input()

## Python Programming in Biology

Python is a versatile and widely used programming language, it can be used in a wide range of applications in biology research. Some of the common uses of Python in biology research include:


__NEED TO WRITE MORE OF THIS !!!__


## Using Jupyter Notebook 

__About Jupyter Notebook:__

Jupyter Notebook is an open-source web-based interactive computing environment that allows you to create, run, and share live code, equations, visualizations, and narrative text in a single document. It supports multiple programming languages, including Python, R, Julia, and many others. Jupyter Notebook is commonly used in data science, machine learning, scientific computing, and other domains for interactive data analysis, visualization, and exploration.

Jupyter Notebook provides a browser-based interface where you can create "notebooks" that contain code cells, text cells, and other types of cells. Code cells allow you to write and execute code directly in the notebook, while text cells allow you to write narrative text, equations, and markdown for documentation and explanations. This combination of code and text cells makes Jupyter Notebook an excellent tool for creating reproducible research, documenting data analysis workflows, and sharing interactive notebooks with others.

__Run your first code:__

Put the cursor in the box below and click the "run" button (or press `shift+enter` on your keyboard) to see the result.

In [6]:
print("Hello, world.")

Hello, world.


The above cell is a code cell. For all code cells, you can see a `In []` displaying before the cell. 

__Text cells:__

You can double click to edit the text cells, and click the "Run" button or press `shift+enter` to display the text cells properly. We're not going to cover how to write formatted texts in Jupyter Notebook in this course, our main purpose is to practise coding. 

__The two status of cells:__

* When you're in the edit mode of a cell, the cell will show a __green__ border. You can edit the contents of the cell.
* When you're in the command mode of a cell, the cell will show a __blue__ border. You can perform various operations on the notebook, such as adding cells , deleting cells , and changing cell types. 

__Shortcuts:__

When you're in the edit mode:
* Press `shift+enter` to run the cell. 
* Press `esc` to go to the command mode.

When you're in the command mode:
* Press `b` to add a new cell.
* Press `d` twice to delete the cell.
* Press `y` to change the cell type to code.
* Press `m` to change the cell type to text.
* Press `enter` to go to the edit mode.

__Exercise 01: play with the above shortcuts, create some new cells, delete some existing cells, and changing between cell types.__

## Data Types and Basic Operations 
### Integers

In Python, integers (int) are a built-in data type that represents whole numbers without any fractional or decimal parts. Integers can be positive, negative, or zero, and can have an unlimited range of values, limited only by the available system memory.

You can define an integer variable in Python by assigning a whole number to a variable name. For example:

In [4]:
x = 5
y = -10

### Floating-point numbers 

Floating-point numbers in Python are a type of numerical data that can represent real numbers with decimal values. They are implemented as a built-in data type called "float" in Python and are used for performing arithmetic operations with real numbers that require precision beyond what can be achieved with integers.

In Python, you can create a floating-point number by simply including a decimal point in a numeric value. For example:

In [7]:
a = 3.14 
b = 2.0 

### Arithmetic operations 

Arithmetic operations can be performed on integers and floating-point numbers in Python using the standard arithmetic operators.

When perform on integers:

In [14]:
# addition 
x + y

-5

In [15]:
# substraction 
x - y

15

In [16]:
# multiplication
x * y

-50

In [17]:
# division 
x / y

-0.5

In [18]:
# exponentiation
x ** y

1.024e-07

__HERE TO CONTINUE !!!__

In [ ]:

x ** y
x % y
x // y

In [8]:
# Addition
sum = a + b
print(sum)  # Output: 5.14

# Subtraction
difference = a - b
print(difference)  # Output: 1.14

# Multiplication
product = a * b
print(product)  # Output: 6.28

# Division
quotient = a / b
print(quotient)  # Output: 1.57

# Exponentiation
power = a ** b
print(power)  # Output: 9.8596

5.140000000000001
1.1400000000000001
6.28
1.57
9.8596


In [9]:
a + b

5.140000000000001

In [ ]:
3+4-5.25

\* for multiply, / for divide

In [ ]:
7/2*5

\*\* for exponentiation (raise to power)

In [ ]:
2**5

Try a few arithmetic expressions of your own below. You can add more cell by clicking on the "+" button or by pressing "Escape", followed by "b". 

When you are typing in the cell, the cell is green. When you press "escape", the cell turns blue, in this status, you can use shortcuts to add or delete cells and change cell types. 

There are two types of numbers: **integers (whole numbers)** and **floating point numbers** (with a fractional part).

- 1, 10 and 42 are integers
- 1.02, 3.14 and 10.0 are floating point numbers

Try multiplying two integer numbers together, and then a integer with a floating number. In each case notice the type of the result.

In [ ]:
2*1.5

In [ ]:
2*4

Multiplication and division have higher precedence than addition and substraction (they happens before addition & substraction).

In [ ]:
2+3*5

Maybe this isn't what you meant. You can use parentheses to give priority to specific operations.

In [ ]:
(2+3)*5

Parentheses in expressions are not always necessary, but they can make expressions clearer.


### **Python 3** has some special operators that get integer results. 

- `//` is integer divide (throw the fractional part away).

- `%` is the modulo operator ... it finds the remainder, this will come in handy for codons later on in the course (think codons and dividing by 3)

- `divmod()` finds both the quotient and the remainder as a pair of values.

For example this expression asks how many times does three fit into 7, throw away the remainder

In [ ]:
7//3

This expression ask ONLY for the remainder

In [ ]:
9%2

This expression gets both the quotient and the remainder

In [ ]:
divmod(9, 2)

You may be thinking, I am never going to use these functions, however...for DNA we often care about multiples of three...

In [ ]:
divmod(1464, 3)

### String expressions

Computing isn't all about numbers. A "value" can be a string of characters. And some of Python's operators can be used to create new strings from old ones.

Here are some examples of string expressions.

In [ ]:
"abc"+"def"

In [ ]:
"abc"*5

Clearly, you can multiply a string by a number.

What happens if you try to add a number to a string? Try it for yourself in the cell below.

In [ ]:
"abc"+5

So, the "+" operator can only concatenate string with string. 

A string can start with either a single or a double quote ... but each string must end with same character as it started with. 

In [ ]:
'abc'

In [ ]:
"abc" + 'def'

This can be handy if you need a string like this ...

In [ ]:
"Fred's thingy."

There is a way to include both types of quotes in a string. The back-slash character sort-of hides the end-of-string meaning of quotes within a string.

In [ ]:
"Fred's \"thingy\" (in double quotes)"

There are also some special characters ... like the newline character (\n) ... that you can put into a string.

In [ ]:
"Line 1.\nLine 2."

We can also represent tabs in this format

In [ ]:
"Column_1.\tColumn_2."

... which doesn't show up properly here, but it will later when we talk about the BUILT-IN functions in Python

## Some more special values

Python has a few other special values: True, False and None.

None is just an "empty" value. There's not much about it.

**True and False are called Booleans.**

There are some *relational opeators* that give either True or False as a result. These include:
- less than: **<**
- greater than: **>**
- equal: **==**
- not equal: **!=**
- less than or equal: **<=**
- greater than or equal: **>=**

Check the results of all the following expressions. Do they all make sense to you?

In [ ]:
2>3

In [ ]:
3+2>=5

In [ ]:
None

In [ ]:
'abc'=="abc"

In [ ]:
not False

In [ ]:
not not True

In [ ]:
True!=True, True==None

In [ ]:
"abc" == 'abc'

### Numbers vs Strings

Numbers and strings are different ... even if they look the same. Python has ways to convert between the types.

In [ ]:
str(100+1)

Do you remember the error you had above when adding a number to a string? Now try this:

In [ ]:
"abc" + str(123)

In [ ]:
int('27')

In [ ]:
float('27.3')

In [ ]:
float('27')

Write code to convert the string "42.0" to an integer.

### Assigning values and strings to objects (variables)
Python is similar to bash and we use the "=" to assign a value to a variable which we can manipulate further

In [ ]:
x=5

Typing the object name works but we can also use the function print

In [ ]:
print(x)

Print is a powerful function because we can very easily add text to our variable.

In [ ]:
print("Gene_X is" + x + "basepairs long")

But we got an error here, why? 

The error message says "can only concatenate str (not int) to str", because our variable x is an integer. We can check the data type of our variable by using the type() function.

In [ ]:
type(x)

So, do we need to transform all of our data to the same data type before using the print() funtion?

No. The print() is a powerful function and it has the ability to concatenate different types of data together.

To do that, write the print function in this way (where we use comma to separate each variable):

In [ ]:
print("Gene_X is", x, "basepairs long.")

### Lets get back to the **\n** (end of line) and **\t** (tab)  

We can get these to display correctly using our fuction print().

Say we want to print three things, the length of our gene (x), the number of amino acids, and a BOOLEAN TRUE/FALSE to tell us if this gene is completely in-frame

In [ ]:
GeneX="atgatcatcgatcgtagctgggctcttag"
len(GeneX)

In [ ]:
x=len(GeneX)

In [ ]:
print(x)

OK, now I want to calculate how many Amino Acids my gene GeneX (represented here as string) has

In [ ]:
len(GeneX)/3

Next, I want to return TRUE if my GeneX has a perfect number of bases (i.e. I am in-frame) OR return FALSE if my gene is out of frame

In [ ]:
len(GeneX)%3

In [ ]:
len(GeneX)%3 == 0

Now I want to put this all together and print it to a tab delimited file. 

I can use my variable x which I created above OR I can just evaluate the expression inside the print() function. Python doesn't really care.

In [ ]:
print("Gene_1\t", x, "\t", x/3, "\t", len(GeneX)%3==0)

In [ ]:
print("Gene_1\t", x, "\t", x//3, "\t", len(GeneX)%3==0)

Essentially the Print function is a very valuable tool that allows you to output data into nice tab delmited files, or alternatively print onto a new line

In [ ]:
print("Gene_1\n", len(GeneX), "\n", len(GeneX)//3, "\n", len(GeneX)%3==0)

Let's make a tab delimited header and add this to our print statement above

In [ ]:
header="Gene_name\tLength\tAA_length\tInFrame\n"
print(header, "Gene_1\t", x, "\t", x//3, "\t", len(GeneX)%3==0)